blueprint/core/blueprint_ImageView.h

```diff
@@ -58,40 +66,81 @@ namespace blueprint
         {
             View::paint(g);
 
-            float opacity = props.getWithDefault("opacity", 1.0f);
+            const float opacity = props.getWithDefault("opacity", 1.0f);
 
             // Without a specified placement, we just draw the drawable.
             if (!props.contains("placement"))
                 return drawable->draw(g, opacity);
 
             // Otherwise we map placement strings to the appropriate flags
-            int flags = props["placement"];
-            juce::RectanglePlacement placement (flags);
+            const int flags = props["placement"];
+            const juce::RectanglePlacement placement (flags);
 
             drawable->drawWithin(g, getLocalBounds().toFloat(), placement, opacity);
         }
 
     private:
         //==============================================================================
-        std::unique_ptr<juce::Drawable> drawable;
+        juce::Image loadImageFromURL(const juce::URL &url) const
+        {
+            if (url.isLocalFile())
+            {
+                const juce::File imageFile = url.getLocalFile();
 
-        //==============================================================================
-        JUCE_DECLARE_NON_COPYABLE_WITH_LEAK_DETECTOR (ImageView)
+                if (!imageFile.existsAsFile())
+                {
+                    const juce::String errorString = "Image file does not exist: " + imageFile.getFullPathName();
+                    throw std::logic_error(errorString.toStdString());
+                }
+
+                juce::Image image = juce::ImageFileFormat::loadFrom(imageFile);
+
+                if (image.isNull())
+                {
+                    const juce::String errorString = "Unable to load image file: " + imageFile.getFullPathName();
+                    throw std::logic_error(errorString.toStdString());
+                }
+
+                return image;
+            }
+
+            if (url.isProbablyAWebsiteURL(url.toString(false)))
+            {
+                //TODO: What approach should we take here?
```

Comment:
Yeaa good thoughts here. I like the ThreadPool idea, thinking we could basically just do something like

```
tp.addJob([this, url]() -> void {
  juce::MemoryBlock mb;
  url.readEntireBinaryStream(mb);

  // Will have to check the URL to see which image format to use I think?
  auto image = juce::ImageFileFormat::loadFrom(mb.data(), mb.size());

  // Will have to grab a mutex here
  drawable = std::move(image):
});
```

Then there's the question of who owns the thread pool? Each `Image` could have its own with 1 thread, but that seems like it would scale poorly. Another option is for ReactAppRoot to carry 1 threadpool (4 threads or something) and for all Image classes to `findParentComponentOfType<ReactAPplicationRoot>()->addThreadPoolJob(...etc)`. If that latter idea is easy enough to pull off, I think that scales well.

Lastly, placeholder– I'm not sure we should do anything. I think the user has enough control here to do what they like, for example:

```
<View>
  <Image source="file:///path/to/placeholder.png" {...styles.image} />
  <Image source="http://www.slowloadingimage.com/x" {...styles.image} />
</View>
```
In this example, the `file://` image will show up immediately, and will show until the network one draws over it. We could extend this by dispatching a `load` event on `Image` when the network request succeeds. This is a known pattern in web (https://idiallo.com/javascript/img-detect-loaded, couldn't find the spec in my brief search), and so the user here could render two images (placeholder + network image), wait for the load event, and when handling the load event remove the placeholder image from its tree:

```
tp.addJob([this, url]() -> void {
  juce::MemoryBlock mb;
  url.readEntireBinaryStream(mb);

  // Will have to check the URL to see which image format to use I think?
  auto image = juce::ImageFileFormat::loadFrom(mb.data(), mb.size());

  juce::MessaageManager::callAsync([this, img = std::move(image)]() {
    drawable = std::move(image):
    if (props.hasProperty("onload") && props["onload"].isNativeFunction()) {
      props["onload"]();
    }
  });
});
```

Reply:
@nick-thompson , Yeah totally, I almost suggested and event/callback on `Image` rather than placeholder prop. I think that's nicer. 

Yeah `ThreadPool` seems the way to go and mutex rather than `MessageManager::callAsync` makes more sense! (I almost never use the juce threading primitives as we have our own monstrous threading system). 

`ReactApplicationRoot` owning the `ThreadPool` seems reasonable. I guess it depends if we envisage other use cases for the pool in future otherwise we could always consider a static `ThreadPool` instance on `Image` as `Image::downloadPool` or something?

If you're happy to merge this now that would be awesome as I'd could use it for some things. I'll tackle the file download in a separate PR (probs on Saturday as in the middle of other things).

Reply:
> If you're happy to merge this now that would be awesome as I'd could use it for some things. I'll tackle the file download in a separate PR (probs on Saturday as in the middle of other things).

Yea I'll go ahead and merge now!

> ReactApplicationRoot owning the ThreadPool seems reasonable. I guess it depends if we envisage other use cases for the pool in future, otherwise we could always consider a static ThreadPool instance on Image as Image::downloadPool or something? I haven't actually checked whether a static ThreadPool is a good idea/safe mind you.

Good question... I'm kind of hesitant to use static instances of anything because I want blueprint to comfortably & safely work within the context of plugins in a DAW. That miiiiight be safe for this particular case anyway, but I guess I've kinda developed the habit of "no static instances anywhere" as a plugin author

> Yeah ThreadPool seems the way to go and mutex rather than MessageManager::callAsync makes more sense! (I almost never use the juce threading primitives as we have our own monstrous threading system).

Yea I like the mutex, but actually if we're going to invoke the `props["onload"]` callback that will have to happen on the message thread anyway. So, might need the `callAsync` either way

Reply:
Ahh yeah wasn't thinking. ThreadPool between multiple instances might not be very smart 😆